doc/internal/man7/deprecation.pod

```diff
@@ -0,0 +1,219 @@
+internal header file, as show in L</Implementations kept for internal
+purposes>.
+
+=back
+
+=head2 General macros
+
+I<Note: none of these macros should be used directly, please use the version
+specific macros instead>
+
+OSSL_DEPRECATED() implements the deprecated attribute if the compiler
+supports it, otherwise it's defined to nothing.  It takes one argument
+I<since> that should be set to the OpenSSL version where the attribute is
+added, and will be displayed with the deprecation warning message, for
+compilers that support user specified deprecation messages.
+
+OSSL_DEPRECATED_FOR() does the same as OSSL_DEPRECATED(), but also takes a
+second argument I<msg>, which is an additional text messages to be displayed
+with the deprecation warning along with the OpenSSL version number, for
+compilers that support user specified deprecation messages.
+
+B<OPENSSL_NO_DEPRECATED> is a macro that's generated by OpenSSL
+configuration in response to the C<no-deprecated> configuration option.
+
+=head2 Version specific macros
+
+OSSL_DEPRECATEDIN_I<major>_I<minor>() macros that are defined to
+OSSL_DEPRECATED() with that version number as I<since>, for any version up
+to and including the one expressed with L<OPENSSL_API_COMPAT(7)>.  For any
+known version above the version expressed with L<OPENSSL_API_COMPAT(7)>,
+The are defined to nothing.
+
+OSSL_DEPRECATEDIN_I<major>_I<minor>_FOR() macros that are defined to
+OSSL_DEPRECATED() with that version number as I<since>, under the same
+conditions as OSSL_DEPRECATEDIN_I<major>_I<minor>().
+
+B<OPENSSL_NO_DEPRECATED_I<major>_I<minor>> macros are defined to be used as
+guards around declarations that should disappear from the public header
+files (F<< <openssl/*.h> >>).
+
+Any set of symbols deprecated with a B<OSSL_DEPRECATEDIN_I<major>_I<minor>>
+attribute macro B<must> be wrapped with a guard using the corresponding
+B<OPENSSL_NO_DEPRECATED_I<major>_I<minor>> macro, see L</Header files>
+example below.  This not only affects what the user of the header file will
+have available, it's also used to determine the conditions for exporting the
+symbol in the shared libraries.
+
+=head1 EXAMPLES
+
+=head2 Header files
+
+In public header files (F<< <openssl/*.h> >>), a deprecated symbol will
+always be wrapped with a negative test of the guard:
+
+ # ifndef OPENSSL_NO_DEPRECATED_1_1_0
+ /* ... */
+ OSSL_DEPRECATEDIN_1_1_0 __owur int
+ HMAC_Init(HMAC_CTX *ctx, const void *key, int len, const EVP_MD *md);
+ /* ... */
+ # endif
+
+=head2 Implementations of deprecated symbols
+
+At least for the period of deprecation for any symbol, its implementation
+needs to be kept, and for symbols that are kept internally, for longer than
+that.  There are two things to deal with:
+
+=over 4
+
+=item Deprecation warnings
+
+To remedy deprecation warnings, simply define the macro
+B<OPENSSL_SUPPRESS_DEPRECATED> at the beginning of the translation unit.
```

Comment:
Would a configuration option be workable to suppress these globally?
`no-deprecation-warnings`

Reply:
I did suggest something like that to @mattcaswell, but he raised the point that control should remain selectively with the translation units that need to do this.  He does have a point.  Besides, the user that wants to avoid those warnings are better served by adding `-Wno-deprecated-declarations` to the configuration, or corresponding option for their compiler.

Reply:
Okay, I didn't realise that there was an option for this already.

Reply:
Er, that's a gcc option...